### PR TITLE
Add Stdlib.Float module

### PR DESCRIPTION
**Pull request by nojb**
Let the floodgates open! 😄 

Some remarks:
- Most names are the same, except that for analogy with the other numeric modules I renamed: `abs_float -> Float.abs`, `mod_float -> Float.rem`, `string_of_float -> Float.to_string`, etc.
- Do we want to deprecate the "old" `Pervasives` float functions?

See also #1010, #964, #944, #1294, #1354.

Comments welcome!

### CODE REVIEW

**Comment by yallop** [comment]
It would be good to have a `Float.Array` submodule like this:

```ocaml
module Array : sig
  type t = floatarray
  external create : int -> t = "caml_floatarray_create"
  external length : t -> int = "%floatarray_length"
  external get : t -> int -> float = "%floatarray_safe_get"
  external set : t -> int -> float -> unit = "%floatarray_safe_set"
  external unsafe_get : t -> int -> float = "%floatarray_unsafe_get"
  external unsafe_set : t -> int -> float -> unit = "%floatarray_unsafe_set"
end
```

and to make `Array.Floatarray` an alias for this module.

See [previous discussion](https://github.com/ocaml/ocaml/pull/1294#discussion_r134680563) under the "Configure-time option for float array optimization" PR.

File: stdlib/float.mli
Comment (alainfrisch) [inline stdlib/float.mli:238]: I think it would be a nice convention if such module related to a specific type would come with the `type t = ....` alias as their first exposed component, perhaps immediately followed by common functions (compare, equal).

Let's also add `Float.hash`.

**Comment by nojb** [comment]
> It would be good to have a Float.Array submodule like this:

This is now done.

File: stdlib/float.mli
Comment (alainfrisch) [inline stdlib/float.mli]: Should these predicates be exposed in Float (it seems they aren't, currently)?  If so, perhaps in a sub-mode `Floats.Ops`?

Also, concerning the equality, I always found it unfortunate that there is no easy way to create a "Leibniz-equality" for floats, which is needed to implement memoization correctly (for instance).  `compare x y = 0` is also good, but it doesn't make the difference  between positive and negative, which behave differently (for x -> 1/x).

Would it be a good time to add such a function (or just the associated total ordering)?

File: stdlib/float.mli
Comment (alainfrisch) [inline stdlib/float.mli]: If would be useful to document where this is `compare x y = 0` or `x = y`.

File: .depend
Comment (alainfrisch) [inline .depend:561]: (Just to be sure: I assume this change is unrelated to this PR, right?)

File: stdlib/float.ml
Comment (yallop) [inline stdlib/float.ml:88]: `Float.equal` and `Pervasives.(=)` will behave differently, since

```
# nan = nan;;
- : bool = false
# Pervasives.compare nan nan;;
- : int = 0
```

If that's the desired behaviour it'd be good to have a comment in the documentation for `Float.equal`

File: stdlib/float.mli
Comment (xclerc) [reply stdlib/float.mli:238]: Well, in modules such as `Int32` or `String` they currently appear at the end.
Perhaps the convention you suggest should be the subject of another PR,
to be applied to all relevant modules.

File: .depend
Comment (nojb) [reply .depend:561]: Yes, I guess it is a side-effect of `make depend`.

**Comment by xclerc** [comment]
Would it make sense to add functions for conversion to/from bits?
(`Int{32,64}.{bits_of_float,float_of_bits})

And also maybe conversion to/from integer types
(`{Nativeint,Int32,Int64}.{of,to}_float`).

File: stdlib/float.mli
Comment (nojb) [reply stdlib/float.mli:238]: > Let's also add Float.hash.

I think this will require that I duplicate the definition of `Hashtbl.hash` to avoid a circular dependency (`Array` -> `Float` -> `Hashtbl` -> `Array`).

File: stdlib/float.mli
Comment (xclerc) [reply stdlib/float.mli]: Could we rely on a another compare function, that would compare
float values by first converting them into int64 ones (bit patterns)?

**Comment by nojb** [comment]
> Would it make sense to add functions for conversion to/from bits?
(`Int{32,64}.{bits_of_float,float_of_bits})
>
> And also maybe conversion to/from integer types
({Nativeint,Int32,Int64}.{of,to}_float).

Possibly, but in the existing numeric modules the conversion functions are only put in one of the sides, not in both (e.g. there is no `Int32.to_int64`), so this is a change that could be the subject of a different PR and that would affect existing numeric modules as well.

**Comment by xclerc** [comment]
Sorry, it wasn't clear, I was actually suggesting to only add
thing to the new `Float` module:

- `Float.to_bits32 : t -> int32`;
- `Float.of_bits32 : int32 -> t`;
- `Float.to_bits64 : t -> int64`;
- `Float.of_bits64 : int64 -> t`;
- `Float.to_int32 : t -> int32`;
- `Float.of_int32 : int32 -> t`;
- `Float.to_int64 : t -> int64`;
- `Float.of_int64 : int64 -> t`;
- `Float.to_nativeint : t -> nativeint`;
- `Float.of_nativeint : nativeint -> t`.

File: stdlib/float.mli
Comment (nojb) [reply stdlib/float.mli]: So far there seem to be 3 notions of comparison which are useful in different situations:

- `Pervasives.compare`: total order, useful for using with `Map.Make`, etc.
- `Pervasives.(<)`: IEEE 754 comparison, can detect if `x` is `nan` with `x <> x`.
- "bitwise", "Leibniz" or "physical"-equality: useful for memoization, etc.

Why not expose all of them? We need to decide on suitable names. And we need to choose one of them for `equal`.

Comment (alainfrisch) [reply stdlib/float.mli]: Note: bitwise would differentiate nan with different encodings, while I'm not sure we can make the difference otherwise (well, except inspecting the bits).

Comment (bluddy) [reply stdlib/float.mli]: I think we want a total order as our main comparison function, and this should be the only kind of comparison exposed cleanly.

For `nan` detection, an `is_nan` function makes more sense to me than needing any kind of comparison. Bitwise comparison is a niche and should have an appropriately specific name.

File: stdlib/float.mli
Comment (alainfrisch) [reply stdlib/float.mli:238]: Hashtbl.hash is a one-line wrapper around a primitive, so the duplication is fine (there are plenty of other cases of such duplication in the stdlib).

And IMO it would also be ok to drop the dependency from Array to Float (by duplicating the current definition of the module, marking Array.Float as deprecated, and only extending Float.Array going forward).

File: stdlib/float.mli
Comment (alainfrisch) [reply stdlib/float.mli]: > I think we want a total order as our main comparison function, and this should be the only kind of comparison exposed cleanly.

The question is which one?  The current one is fine for many uses, but it collapses -0. and 0. which makes it dangerous to use in some cases.  But a refined version with -0. < 0. would not be suitable in all cases either.  Both are total.

For nan detection, we already have classify_float (or `x = x`).

**Comment by alainfrisch** [comment]
Cf also #1354.

File: stdlib/float.mli
Comment (gasche) [reply stdlib/float.mli]: I'm happy to see more orders exposed and documented, but there is no doubt in my mind that `(Float.compare x y)` must have the exact same semantics as `(Pervasives.compare x y)` when `x, y` are floats.

**Comment by alainfrisch** [comment]
So many nice other PRs depends on this new Float module; we don't want to delay it too much.

So let's keep the discussed additions (conversions with int32/int64; other ordering/equality; etc) for later, and focus instead on polishing the existing PR (documentation, dropping dependency between Float and Array, etc).

**Comment by nojb** [comment]
I removed the dependency between `Array` and `Float`, added a note to the doc of `Float.equal` specifying that it uses `compare` as comparison function and added `Float.hash`.

I also rebased to avoid cluttering the commit history with `make depend` and bootstrapping commits.

File: stdlib/float.ml
Comment (alainfrisch) [inline stdlib/float.ml]: (Later, it could be useful to create an ad hoc runtime primitive for Float.hash, with an unboxed form.)

File: stdlib/float.mli
Comment (alainfrisch) [inline stdlib/float.mli:5]: Please change the line above!

File: stdlib/float.mli
Comment (alainfrisch) [inline stdlib/float.mli:110]: Shouldn't this be Stdlib.fpclass?

**Comment by alainfrisch** [comment]
LGTM (except perhaps the minor comment about Pervasives vs Stdlib).  We can always add components or refine the doc later.

I'll merge soon if nobody objects.

File: stdlib/float.mli
Comment (nojb) [reply stdlib/float.mli:110]: I tried, but got into trouble with `ocamldoc`:
```
../../byterun/ocamlrun ../../ocamlc -nostdlib -nopervasives -I  -c -open Pervasives float.mli
File "float.mli", line 109, characters 15-29:
Error: Unbound module Stdlib
Makefile:17: recipe for target 'float.cmi' failed
make[4]: *** [float.cmi] Error 2
make[4]: Leaving directory '/home/nojebar/ocaml/ocamldoc/stdlib_non_prefixed'
Makefile.unprefix:109: recipe for target '../ocamldoc/stdlib_non_prefixed/pervasives.cmi' failed
make[3]: *** [../ocamldoc/stdlib_non_prefixed/pervasives.cmi] Error 2
```
Any ideas how to fix this? @diml ?

Comment (ghost) [reply stdlib/float.mli:110]: The `-I -c` is odd, it might be because of `-I $(HERE)` in the Makefile, not sure where this is defined. @Octachron I believe you wrote this line, do you know where `$(HERE)` is defined?

Comment (Octachron) [reply stdlib/float.mli:110]: Nowhere. The whole flag `-I $(HERE)` should be removed. Nevertheless, this does not affect the issue at hand: ocamldoc simply knows nothing about a `Stdlib` when building the documentation and is only aware of the extracted Pervasives module. Moreover, ocamldoc limited support of `module type of` means that it would not be able to link to Stdlib.fpclass even after building the Stdlib documentation.

**Comment by Octachron** [comment]
A minor point: the new module is missing an entry in the documentation `manual/manual/libref/stdlib.etex`, should I take care of this point after the merge?

**Comment by nojb** [comment]
> A minor point: the new module is missing an entry in the documentation manual/manual/libref/stdlib.etex, should I take care of this point after the merge?

I will add one, thanks!

File: stdlib/float.mli
Comment (ghost) [reply stdlib/float.mli:110]: Hmm, indeed. I don't see a simple solution to this problem, we could `sed s/Stdlib/Pervasives/` when copying the mli files. Thought it's slightly not satisfactory that the documentation will mention pervasives rather than stdlib

Comment (alainfrisch) [reply stdlib/float.mli:110]: Ok, so we keep Pervasives, and people who insist on deprecating this module will find a solution ;)

File: manual/manual/library/stdlib.etex
Comment (Octachron) [inline manual/manual/library/stdlib.etex:65]: You missed the index part starting from line 100 for the html version of the manual, and line 147 for the latex version.

Comment (nojb) [reply manual/manual/library/stdlib.etex:65]: Oops, thanks! Should be fixed now.

Comment (Octachron) [reply manual/manual/library/stdlib.etex:65]: The manual looks good now, thanks!

**Review by alainfrisch — APPROVED** [review]

File: stdlib/float.mli
Comment (dbuenzli) [inline stdlib/float.mli:80]: I think it would be nice to add a constant [`max_exact_int`](http://erratique.ch/software/gg/doc/Gg.Float.html#VALmax_int_arith) and indicate that integers in the range `[-max_exact_int;max_exact_int]` will be represented exactly. This constant is also useful e.g. when you serialize integers to json numbers.

**Review by dbuenzli — COMMENTED** [review]
It seems that the PR left adding interesting constants like `pi` or `e`, also signature-wise it could be made more IntX like (e.g. adding `zero`, `one`).

**Comment by alainfrisch** [comment]
> It seems that the PR left adding interesting constants like pi or e, also signature-wise it could be made more IntX like (e.g. adding zero, one).

Indeed. About pi, there is #964.  The goal here is to make Float enter Stdlib rather quickly, without real additions compared to existing features (to make the PR consensual) in order to unlock many other PRs that are pending on it.

File: stdlib/float.mli
Comment (alainfrisch) [reply stdlib/float.mli:80]: I agree, but perhaps for a future addition.  As simple as such an addition seems, there will be space for discussion, such as:

  - The suggested naming is confusing since of course this is not the maximum integer represented as a float (that would be `max_float`).

  - What would be the type of this value?  float? int? int64?

File: stdlib/float.mli
Comment (alainfrisch) [inline stdlib/float.mli]: About the naming of these constants:

- Should that be just named `epsilon`?

- For `max_float`/`min_float`, one could use `max_value` (and later add Int32.max_value as an alias for Int32.max_int, etc).

Comment (nojb) [reply stdlib/float.mli]: I don't have a strong opinion on `epsilon_float`. For the others, I would leave it for a different PR that addresses all numeric modules at once.

Comment (alainfrisch) [reply stdlib/float.mli]: (Also, min_value is really a bad name for min_float.)

File: manual/manual/library/stdlib.etex
Comment (Chris00) [reply manual/manual/library/stdlib.etex:65]: Shouldn't it be explicitly said that it is _Double precision floating-point numbers_?

Comment (nojb) [reply manual/manual/library/stdlib.etex:65]: Maybe. In http://caml.inria.fr/pub/docs/manual-ocaml/core.html#sec547, `type float` is described as being the type of "floating-point numbers" though.

File: stdlib/float.mli
Comment (Chris00) [inline stdlib/float.mli:45]: I think we should also add prefix (`~+`, `~-`) and infix operators (`+.`, `-.`, `*.`, `/.`, `**`).  The idea is that if these operators are redefined by another module, one can locally open this one to recover the usual behavior.

File: stdlib/float.mli
Comment (Chris00) [reply stdlib/float.mli]: I also think it should simply be `epsilon`.

Comment (Chris00) [reply stdlib/float.mli]: What about `smallest` or `lowest` or the more explicit `smallest_pos` instead of `min_float`?

File: stdlib/float.mli
Comment (Chris00) [inline stdlib/float.mli:249]: I think specialized versions of `min` and `max` should be provided (clearly describing their behavior on NaN — maybe several versions are required: one that ignores NaN, another one that returns NaN as soon as one is present).

File: stdlib/float.mli
Comment (Chris00) [inline stdlib/float.mli:67]: **Fast** functions such as `is_nan` and `is_finite` would be nice.

File: stdlib/float.mli
Comment (nojb) [reply stdlib/float.mli]: OK, I renamed it `epsilon`.

File: manual/manual/library/stdlib.etex
Comment (Chris00) [reply manual/manual/library/stdlib.etex:65]: Sure but there is a chance that people find the module `Float` before the `type float`.

File: stdlib/float.mli
Comment (nojb) [reply stdlib/float.mli:45]: Definitely something that needs to be considered - but I think it is better to do it in a different PR that addresses all numeric modules at the same time.

BTW, wouldn't it make sense to use (+) instead of (+.) now that we have a separate module for it ?

**Comment by nojb** [comment]
@Chris00 Thanks for all the good suggestions; I suggest we address any **new** functionality in a different PR in order to keep this one as consensual as possible so it can be merged quickly (and unblock several other PRs which depend on this one).

File: stdlib/float.mli
Comment (dbuenzli) [reply stdlib/float.mli:80]: Yes I was not convinced by the name I gave in `Gg.Float` and tried `max_exact_int` which is silly. Regarding the type of the value it can't be `int` (which can be 31-bits) and I don't really see why this should be an `int64`, it's a particular value of that type, you can convert it to something else if you need.

**Comment by dbuenzli** [comment]
> Indeed. About pi, there is #964. The goal here is to make Float enter Stdlib rather quickly, without real additions compared to existing features (to make the PR consensual) in order to unlock many other PRs that are pending on it.

Fair enough but please make the first version of `Float` that gets in an OCaml release a complete design of the `Float` module. Micro-additions to the stdlib are very irritating and increase the chances of misdesigns.

**Comment by damiendoligez** [comment]
> mod_float -> Float.rem

`mod_float` had the advantage of being reminiscent the C name of the function (`fmod`).

**Comment by nojb** [comment]
> mod_float had the advantage of being reminiscent the C name of the function (fmod).

I agree; shall I rename it back ?

File: manual/manual/library/stdlib.etex
Comment (xavierleroy) [inline manual/manual/library/stdlib.etex:115]: Inconsistent capitalization: `: Floating` and `: Ephemerons` but `: operations` and `: pretty-printing`.

File: stdlib/float.mli
Comment (xavierleroy) [reply stdlib/float.mli:249]: The documentation for [equal] is obscure.  What about:
```
(** [equal x y] compares [x] and [y] for equality.  Unlike standard equality on floating-point numbers, [equal] treats [nan] as equal to itself and different from any other floating-point value.  This treatment of [nan] ensures that [equal] defines an equivalence relation.  [equal x y] is equivalent to [compare x y = 0].  *)
```

**Review by xavierleroy — COMMENTED** [review]
Some documentation suggestions below.

I support @alainfrisch's suggestion to commit a first version to trunk, then add new constants and functions via separate pull requests.

**Comment by bluddy** [comment]
>> Fair enough but please make the first version of Float that gets in an OCaml release a complete design of the Float module. Micro-additions to the stdlib are very irritating and increase the chances of misdesigns.

I think we want to apply different approaches to different modules. Float is a module that is needed specifically for OCaml because of its limitations, and it has therefore been replicated in every alternative stdlib. We know that we want it, we know more or less what's going to be in it, and I think it's ok to let it evolve as we go.

Regarding other modules, it would be nice to see a proper design document first. For example, many of the functions in Pervasives (Stdlib) belong in a Math module: (sin, tan, tanh etc). The conversion functions (*_of_string, string_of_*) can be moved to individual modules. Same with the customized print functions. The Unix module is a mess, and parts should be moved out to a File module. Other parts should be moved to a Networking or Socket module, etc.

**Comment by nojb** [comment]
Thanks @xavierleroy for the suggestions; I fixed the capitalization in the manual and used your suggestion for the doc of `equal`.

File: stdlib/float.mli
Comment (mookid) [reply stdlib/float.mli]: a nice thing about epsilon_float is the similarity with `float.h` (where it is DBL_EPSILON).

Comment (Chris00) [reply stdlib/float.mli]: `Float.epsilon` is as similar — moreover, I do not think we aim to mimic C.

**Comment by Chris00** [comment]
> >  mod_float -> Float.rem
>
> mod_float had the advantage of being reminiscent the C name of the function (fmod).

And why not `Float.(mod)`? (That will become an infix operator but I think this is nice.)

File: stdlib/float.mli
Comment (Chris00) [reply stdlib/float.mli:45]: We could but it would make heavier to mix integer and floating points in the same expression — a situation that occurs quite often.

**Comment by alainfrisch** [comment]
> mod_float had the advantage of being reminiscent the C name of the function (fmod)

Are there other functions with a direct counterpart in C numerical functions, but with a different name?
If not, then what about `fmod`?

**Comment by nojb** [comment]
> Are there other functions with a direct counterpart in C numerical functions, but with a different name?
If not, then what about fmod?

Off the top of my head, only `Float.abs -> fabs`. Using `fmod`/`fabs` is fine by me.

**Comment by alainfrisch** [comment]
Well, at least, "standard C mathematical functions have the same name here" would be an explicit design guideline, possibly reducing the amount of discussion for future additions.  It might introduce some useless noise (abs -> fabs as you mention, or a future max -> fmax), though.  A variant would be "standard C mathematical function have the same name here, except for the 'f' which is dropped when this doesn't yield a reserved keyword".  Opinions?

Btw, I can also see `classify_float `-> `fpclassify`.

(Ref: https://en.wikipedia.org/wiki/C_mathematical_functions)

**Comment by nojb** [comment]
Note that naming discussion is constrained by the fact that `Complex` already exists and mostly follows your second suggestion (dropping prefix letter).

**Comment by Chris00** [comment]
I would drop the `f` since this is due to the poor man module system (hum) of C.  For `classify_float`, I would prefer `Float.classify`.

File: stdlib/float.ml
Comment (Chris00) [inline stdlib/float.ml:99]: Shouldn't there be more operations such as `blit`, `append`,...?

**Comment by bluddy** [comment]
>> I would drop the f since this is due to the poor man module system (hum) of C. For classify_float, I would prefer Float.classify.

Exactly. Trying to match C is silly -- they're dealing with a global namespace. The whole point here is that people can automatically guess the names of functions they need without looking them up because they're the same name as non-float functions, except in the Float module. They need to think about what the function does -- not about arcane prefixes or suffixes.

File: stdlib/float.ml
Comment (alainfrisch) [reply stdlib/float.ml:99]: The current version is just a copy of the previous Array.Floatarray.  We can extend it later.

**Comment by alainfrisch** [comment]
I suggest the following set of changes:

  - classify_float -> classify
  - drop the prefixes for Float.fpclass (Normal|Subnormal|Zero|Infinite|Nan) 
  - drop modf: the naming is not so clear, and one might want to return a record with 2 (unboxed) floats instead for performance reasons; let's keep that for later
  - drop rem/fmod, unless a consensus emerges on the naming

**Comment by alainfrisch** [comment]
The current state LGTM.

Does anyone want to suggest other changes?  (Additions can be left for later.)

**Review by Chris00 — APPROVED** [review]

**Comment by alainfrisch** [comment]
@nojb Let's merge this one!  Can you rebase/fix the conflicts?

**Comment by nojb** [comment]
> Can you rebase/fix the conflicts?

Done.

**Comment by alainfrisch** [comment]
Ok, hopefully the CI will finish before the next conflict pops up.

**Comment by trefis** [comment]
Just out of curiosity: why is a bootstrap needed for this change?

**Comment by nojb** [comment]
> Just out of curiosity: why is a bootstrap needed for this change?

Good catch, it is not! I guess it was needed at some intermediate stage of the PR (or at least I thought it was) and it was left there after that.

I have removed the bootstrap commit.

**Pull request merged**
